pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -200,32 +234,44 @@ func (p *testPlanner) initUpgradeSteps() []testStep {
 // upgraded. It will wait for the cluster version on all nodes to be
 // the same and then run any after-finalization hooks the user may
 // have provided.
-func (p *testPlanner) afterUpgradeSteps(fromVersion, toVersion *clusterupgrade.Version) []testStep {
+func (p *testPlanner) afterUpgradeSteps(
+	fromVersion, toVersion *clusterupgrade.Version, scheduleHooks bool,
+) []testStep {
 	p.currentContext.Finalizing = false
 	p.currentContext.Stage = AfterUpgradeFinalizedStage
-	return p.hooks.AfterUpgradeFinalizedSteps(p.currentContext)
+	if scheduleHooks {
+		return p.hooks.AfterUpgradeFinalizedSteps(p.currentContext)
+	}
+
+	// Currently, we only schedule user-provided hooks after the upgrade
+	// is finalized; if we are not scheduling hooks, return a nil slice.
+	return nil
 }
 
 func (p *testPlanner) upgradeSteps(
-	stage UpgradeStage, from, to *clusterupgrade.Version,
+	stage UpgradeStage, from, to *clusterupgrade.Version, scheduleHooks bool,
 ) []testStep {
 	p.currentContext.Stage = stage
 	msg := fmt.Sprintf("upgrade nodes %v from %q to %q", p.crdbNodes, from.String(), to.String())
-	return p.changeVersionSteps(from, to, msg)
+	return p.changeVersionSteps(from, to, msg, scheduleHooks)
 }
 
-func (p *testPlanner) downgradeSteps(from, to *clusterupgrade.Version) []testStep {
+func (p *testPlanner) downgradeSteps(
+	from, to *clusterupgrade.Version, scheduleHooks bool,
+) []testStep {
 	p.currentContext.Stage = RollbackUpgradeStage
 	msg := fmt.Sprintf("downgrade nodes %v from %q to %q", p.crdbNodes, from.String(), to.String())
-	return p.changeVersionSteps(from, to, msg)
+	return p.changeVersionSteps(from, to, msg, scheduleHooks)
 }
 
 // changeVersionSteps returns the sequence of steps to be performed
 // when we are changing the version of cockroach in the nodes of a
 // cluster (either upgrading to a new binary, or downgrading to a
-// predecessor version).
+// predecessor version). If `scheduleHooks` is true, user-provided
+// mixed-version hooks will be scheduled randomly during the
+// upgrade/downgrade process.
 func (p *testPlanner) changeVersionSteps(
-	from, to *clusterupgrade.Version, label string,
+	from, to *clusterupgrade.Version, label string, scheduleHooks bool,
```

Comment:
Nit: maybe `addHooks`, since there isn't any explicit scheduling here, we're merely adding them as (planned) steps.

Reply:
I'm not a huge fan of `addHooks` because `add` is a bit generic. I understand that this is not scheduling in the OS sense (moving something from ready to "run"); `schedule` here was meant more according to this definition:

> to plan something for a particular time
> 
> https://dictionary.cambridge.org/dictionary/english/schedule

which I would argue is what this function is doing.

The planner is explicitly designed to not run anything, so hopefully "schedule" isn't too confusing. That said, I was trying to come up with a better name and nothing felt much better (`planHooks` was the best I could come up with). Let me know if you would still prefer that, I can make the change.

Reply:
Yep, it's fine as is, plus it's internal documentation. So, I should really RTFM :)